### PR TITLE
v3.0.0 (beta -> staging)

### PR DESCRIPTION
**Pull request by endofline**
guidance:
* branch is sufficiently stable for testing
* advise caution on production servers
* `pip3 install -r requirements.txt`

v3.0.0-beta general changelog
* locale updated with latest translations
* compatibility with hangups master
  * `hangups_shim` as permanent drop-in replacement for legacy `hangups` import
  * fix: protobuf-based Annotation for reprocessor 
  * full backward-compatibility with v2 `memory.json`, `config.json`
* plugin compatibility:
  * basic
  * convid
  * api
  * botalive
  * convtools
  * convtools_invitations
  * easteregg
  * humor_hangoutcalls
  * image_memegenerator
  * image_screenshot
  * namelock
  * slack
* unittest compatibility:
  * geticon
  * reprocessor

### CODE REVIEW

File: hangupsbot/plugins/__init__.py [hunk 1]
```diff
@@ -349,6 +350,10 @@ def load(bot, module_path, module_name=None):
         logger.exception("EXCEPTION during plugin import: {}".format(module_path))
         return
 
+    if hasattr(sys.modules[module_path], 'hangups'):
+        logger.info("{} has legacy hangups reference".format(module_name))
+        setattr(sys.modules[module_path], 'hangups', hangups_shim)
```
Comment (nylonee) [inline hangupsbot/plugins/__init__.py]: This is the most beautiful line of python I have ever seen

**Comment by OllieTerrance** [comment]
The new SlackRTM appears not to process messages sent privately to the bot?  Requesting `@bot help` in a channel sends the command list via DM, but I can't use any of the commands in there.

**Comment by OllieTerrance** [comment]
Looks like the new SlackRTM chatbridge has a few differences from the old SlackRTM -- not sure if these count as regressions, intentional changes, or just not yet implemented:

* Messages sent privately to the bot don't appear to be processed.  Requesting `@bot help` in a channel sends the command list via DM, but I can't use any of the commands in there.
* Messages from a HO are shown in Slack by the user's nickname if set.  Previously it showed their full name in any case (more consistent imo).  If this is intentional, perhaps we could add a `showhorealnames` option similar to the Slack counterpart?
* A null `slacktag` uses the channel name rather than omitting the tag on the HO end (null `hotag` works as expected).

**Comment by endofline** [comment]
@OllieTerrance, thanks for the report, will look into these (and also inviting PRs to fix):
* 1 is a bug, and this is the first time I'm receiving a clear report of it, will look into this in a bit
* 2 and 3 is related to the way chat-bridges interact with each other, original slackrtm makes some very particular assumptions about the way its own internal implementation of hotag and slacktag work, I was not even aware `null` was a possible value, thought the range of possible stored values were only `false` or some text

**Comment by OllieTerrance** [comment]
Having played with solutions for 2, something like this would probably work best for me:

```diff
diff --git a/hangupsbot/webbridge/__init__.py b/hangupsbot/webbridge/__init__.py
index c0f0a74..cf052fa 100644
--- a/hangupsbot/webbridge/__init__.py
+++ b/hangupsbot/webbridge/__init__.py
@@ -302,8 +302,12 @@ class WebFramework:
                 photo_url = "https:" + photo_url

-        if nickname:
-            preferred_name = nickname
-        else:
-            preferred_name = full_name
+        if full_name:
+            preferred_name = full_name
+            if nickname and not full_name == nickname:
+                preferred_name = "{} ({})".format(preferred_name, nickname)
+        else:
+            preferred_name = nickname

         if not chat_id:
```

That gives you `Full Name (nickname)` if both are present and different, else `Full Name` or `nickname` depending on which is available.  Happy to PR it, though this is somewhat opinion-based.  🙂

---

For 3, setting `slacktag` in memory to `false` works as expected.  Using `/bot slack_setslacktag <team> <channel> none` sets it back to `null` though, which causes the channel name to be relayed as the tag.

**Comment by endofline** [comment]
@OllieTerrance - as you have surmised, this might be an opinionated implementation ;p rather than modify the default behaviour for all chatbridge plugins and also keeping the output as straightforward as possible (in particular, for those running with chat titles switched on will see "**Null Pointer (endofline) (channel title)**"). here are some alternate suggestions:

**option 1:** override the original webbridge behaviour/method of `_get_user_details` from `slackrtm/bridgeinstance.py` and seeing whether this is suitable
OR
**option 2:** write additional code to make the behaviour toggleable in either webbridge or slackrtm, as per the `showhorealnames` counterpart you mentioned earlier

**Comment by OllieTerrance** [comment]
Option 2 via SlackRTM implemented in #812.  🙂

**Comment by OllieTerrance** [comment]
Another minor observation: Slack provides a plain-text summary for file uploads, but the bot is providing its own one too.

> **terrance**: terrance uploaded a file: Image and commented: It's an image.
> [image]
> **terrance**: shared media from slack

**Comment by endofline** [comment]
@OllieTerrance - on the bot providing its own message, that is an expected behaviour for chatbridges for reasons of consistency across multiple relays. When attaching images, hangouts always shows media before the text (whereas telegram displays media after the text) - this leaves a somewhat unexpected blank line after image uploads on hangouts (but looks OK in telegram), so we might as well fill it in with something at least nominally informative.

**Comment by bing0USMC** [comment]
updated humor_hangouts plugin to include instructions on ways to prevent accidental call in the future.. 


[humor_hangoutcalls.txt](https://github.com/hangoutsbot/hangoutsbot/files/894967/humor_hangoutcalls.txt)

**Comment by endofline** [comment]
@bing0USMC - i understand that this is useful, but this isn't the way to go about it. as i see it, you have 2 options:
* create a proper pull request with mergeable code, relevant links:
  * steps 2-4 of https://guides.github.com/activities/hello-world/
  * https://www.thinkful.com/learn/github-pull-request-tutorial/#Feel-Free-to-Ask
* create an issue on the tracker

please avoid adding messages here unless its an observation/enhancement/bug/comment specific to v3, and use appropriate channels, which is usually an issue or an actual pull request

**Comment by OllieTerrance** [comment]
Having just tried a HO ↔ HO ↔ Slack setup, it looks like the chosen Slack tag isn't preserved in the far hangout (we get the channel name instead).

I was toying with improving the HO representation of Slack messages (adding labels like _edited_, adding better formatting for actions e.g. join, part, `/me`, sending images in a single message with Slack's text) -- I have a PoC, but by implementing them in SlackRTM, they also won't survive multiple relays.

It would perhaps be nice to have some additional, standard flags we can include across the web-bridge, which each plugin can set for their messages, and render for everyone else's.  For Slack, I've got edit and action flags; Hangouts also has actions, though it requires digging into the `hangups.ConversationEvent`<sup>1</sup> -- not sure what Telegram provides, but I'm guessing edits and actions at least?

By doing this, I think #812 would also be better suited at the web-bridge level.  🤔

<sup>1</sup> `any(a.type == 4 for a in event._event.chat_message.annotation)`

**Comment by OllieTerrance** [comment]
Having left this to simmer a bit, a slightly more radical approach for source title sync would be to have modes that apply across the web-bridge:

1. don't show any source titles (for simple A ↔ B syncs, a title often isn't needed)
2. show a fixed title regardless of origin
    * effectively what `slacktag` and `hotag` provide in SlackRTM, but applied globally
    * this would apply to any third-party messages -- e.g. in A ↔ B ↔ C syncs, participants in A would see the same tag for B and C messages
3. show the conversation title 
    * up to individual bridge plugins to provide this -- e.g. for Hangouts, it would be the HO title; for Slack, it could be `#channel`, or `Team Name: #channel`
    * individual plugins could provide their own options to customise this per-sync or globally

I'm not sure about the usefulness of 2, as it feels like it's just there to separate first- and third-party messages, which imo is obvious enough without an explicit label.

---

The config for this may be quite tricky if we want it non-globally...  are there plans to unify the chatbridge config?  (That would of course break v2 compatibility unless migrations are in place.)

```js
"chatbridge": [
  {
    "titles": true // false for none (1)
                   // explicit string for global (2)
                   // true for per-conversation (3)
    "rooms": [
      {
        "type": "hangout", // passed to syncrooms
        "id": "Ugx..."
      },
      {
        "type": "slackrtm", // passed to slackrtm
        "team": "Team", // team/token still configured in under "slackrtm" key
        "id": "C0..."
      }
    ]
  }
]
```

---

These may warrant GitHub issues of their own...

**Comment by endofline** [comment]
@OllieTerrance

**TLDR** ;p

* prioritise functional equivalency, basic message fidelity, non-breaking changes
* move discussion into dedicated issue
* consider configuration mutation ([example](https://github.com/hangoutsbot/hangoutsbot/blob/v3.0/hangupsbot/plugins/slackrtm/bridgeinstance.py#L44)) instead of restructuring (and breaking) configurations
* consider [`chatbridge` key](https://github.com/hangoutsbot/hangoutsbot/blob/v3.0/hangupsbot/webbridge/__init__.py#L250) as starting point for external meta-status

**detailed notes**

* first v3 release will be aimed at functional equivalency of the original plugins, so we probably won't take any action that drastically changes and breaks existing v2 configurations just yet. this will definitely result in some cross-relay features not making it into the initial release because of sheer diversity of implementations.
  * further clarification: one of my goals at the outset was to maintain some resemblance to the original plugins' code and logic across refactors so that the original contributors to various plugins such as telesync and slackrtm can still easily submit improvements and enhancements.
  * nevertheless, you can still add extra configuration keys for advanced features since all chatbridges mutate existing configurations into a useable form that somewhat resembles your [proposed structure](https://github.com/hangoutsbot/hangoutsbot/blob/v3.0/hangupsbot/webbridge/__init__.py#L68). it is conceivable that for plugins that never supported advanced features like syncrooms, you could always add extra keys/config by overriding the chatbridge `load_configuration` and `applicable_configuration` "methods" to check for the existence of additional keys in a portable way e.g. `sync_rooms.extras`, etc - although i loathe to cite this as an example, the [bridgeinstance in slackrtm](https://github.com/hangoutsbot/hangoutsbot/blob/v3.0/hangupsbot/plugins/slackrtm/bridgeinstance.py) does this in a particularly convoluted way due to its rather unique object hierachy and demonstrates how versatile the mutation can be to force a specific configuration format for compatibility reasons.
* this is just the beginning of the standardisation process for cross-chat relays, so further ideas are welcome. i second the idea of moving our discussion into a new issue related with chatbridges itself, otherwise these comments/discussions may get lost or mixed-up with other issues being reported to v3. after i'm done with some RL things, I'll review the poc. 
* in the long-term I'm also concerned with the slackrtm plugin and want to advise caution on the dangers of feature creep and the fact that slackrtm still retains at least a few months of technical debt that we all have to catch up on if we want to rewrite (or add to) its internal logic. having had the wonderful experience of dealing with the original base code, i am eager not to repeat this experience ;P
  * despite its seeming popularity, slackrtm doesn't seem to have a clear plugin owner and this too also needs to be addressed as I'm not actually a slackrtm user, and rather content to use the basic slack plugin for my local community needs
  * changes will also have to take into account specific interactions with telegram as well as future plugins such as the discord plugin, which needs some work for chatbridge compatibility - this greatly complicates the testing and I rather we prioritise basic chatbridge functionality first
* consider experimenting with the [`chatbridge` key](https://github.com/hangoutsbot/hangoutsbot/blob/v3.0/hangupsbot/webbridge/__init__.py#L250) on the "send to hangouts" part of chatbridge - there is no mechanism for optional parameters yet, but its most likely you can experiment with this metadata and consider it for your "standard flag" idea, which all other relay plugins can send additional data back into hangouts e.g. a recently added [`source_gid`](https://github.com/hangoutsbot/hangoutsbot/blob/v3.0/hangupsbot/webbridge/__init__.py#L232) key which is not in use anywhere else but slackrtm, and serves no purpose at the moment.
  * additionally, since the above point only addresses incoming messages to hangouts, I do concur that some standard non-messaging handlers, such as membership changes and annotations should eventually be a part of chatbridges rather than implemented on a separate per-plugin basis.

**Comment by OllieTerrance** [comment]
All fair points.  I've moved discussion of extending chatbridges to #821.  As a heavy slackrtm user, I may be interested in maintaining / refactoring / rewriting.  🙂

File: hangupsbot/plugins/syncrooms.py [hunk 2]
```diff
@@ -54,142 +61,180 @@ def _migrate_syncroom_v1(bot):
...
+                relay_id,
+                message = "{}: {}".format(event.user.full_name, message),
+                image_id = image_id,
+                context = { "passthru": passthru })
+
+
+def _format_source(bot, user_id):
+    user = bot.get_hangups_user(user_id)
+
+    link = 'https://plus.google.com/u/0/{}/about'.format(user.id_.chat_id)
```
Comment (nylonee) [inline hangupsbot/plugins/syncrooms.py]: Not being used.
Is there no longer a way to parse links in hangouts?

File: hangupsbot/handlers.py [hunk 3]
```diff
@@ -47,23 +47,84 @@ def __init__(self, bot, bot_command='/bot'):
                              {},
                              forgiving=True )
 
-    def register_handler(self, function, type="message", priority=50):
-        """registers extra event handlers"""
+    def register_handler(self, function, type="message", priority=50, extra_metadata={}):
```
Comment (OllieTerrance) [inline hangupsbot/handlers.py]: Mutable default argument for `extra_metadata` will be shared across function calls.

Currently this _shouldn't_ bite us -- we only store an original reference to it during a late registration, but a non-default should always be supplied for that -- but probably better to be explicit, e.g. `extra_metadata=None` argument coupled with `extra_metadata = extra_metadata or {}` or similar.

Comment (pleasantone) [inline hangupsbot/handlers.py]: Agreed, it's a horrible practice to ever use foo=(anything mutable) in a function definition as a default.  Python gets very screwed, and they have a justification for it, but it's just one of those things that will eventually bite you in the ass.

Instead of:
```
def func(foo={}):
    pass
```

I always do:

```
def func(foo=None):
    foo = foo or {}
    pass
```

Comment (endofline) [inline hangupsbot/handlers.py]: oops, yeah you guys are right about this, i'll correct this

**Comment by Terrance** [comment]
Probably about time we bite the bullet and get this merged -- nobody's maintaining v2, so it really shouldn't be offered as default any more.  The few issues noted here have been present for a long time but that hasn't stopped people from using it, and documentation can be updated ad-hoc.
